src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java

```diff
@@ -0,0 +1,135 @@
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterMode;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.logic.common.ActivateEvent;
+import org.terasology.math.SideBitFlag;
+import org.terasology.minecarts.blocks.RailComponent;
+import org.terasology.minecarts.blocks.RailsUpdateFamily;
+import org.terasology.registry.CoreRegistry;
+import org.terasology.registry.In;
+import org.terasology.signalling.components.SignalConsumerStatusComponent;
+import org.terasology.world.BlockEntityRegistry;
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.BlockComponent;
+import org.terasology.math.geom.Vector3i;
+import org.terasology.world.block.BlockManager;
+
+/**
+ * Class describing railSwitch's behavior.
+ * @author Aleksander Wójtowicz <anuar2k@outlook.com>
+ * Code below is inspired by Marcin Sciesinki's Signalling module and michaelpollind's Rails module
+ */
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class RailSwitchAction extends BaseComponentSystem {
+
+    @In
+    private WorldProvider worldProvider;
+    @In
+    private BlockEntityRegistry blockEntityRegistry;
+
+    private Block railSwitchLeverOff;
+    private Block railSwitchLeverOn;
+    private Block railSwitchSignalOff;
+    private Block railSwitchSignalOn;
+    private RailsUpdateFamily railFamily;
+    private RailsUpdateFamily invertFamily;
+
+    /**
+     * Prepares blocks for future use (see code below)
+     * These block are obtained from an instance of {@link org.terasology.world.block.BlockManager}
+     */
+    @Override
+    public void initialise() {
+        final BlockManager blockManager = CoreRegistry.get(BlockManager.class);
+        railSwitchLeverOff = blockManager.getBlock("AdditionalRails:railSwitchLeverOff");
+        railSwitchLeverOn = blockManager.getBlock("AdditionalRails:railSwitchLeverOn");
+        railSwitchSignalOff = blockManager.getBlock("AdditionalRails:railSwitchSignalOff");
+        railSwitchSignalOn = blockManager.getBlock("AdditionalRails:railSwitchSignalOn");
+        railFamily = (RailsUpdateFamily)blockManager.getBlockFamily("Rails:rails");
+        invertFamily = (RailsUpdateFamily)blockManager.getBlockFamily("Rails:railsTBlockInverted");
+    }
+
+    /**
+     * Updates railSwitchLever and rails around it, when user interacts the block.
+     * @param event ActivateEvent, just used to differentiate from other events
+     * @param entity The entity related to railSwitchLever
+     */
+    @ReceiveEvent(components = {BlockComponent.class, RailSwitchLeverComponent.class})
+    public void railSwitchLeverAction(ActivateEvent event, EntityRef entity) {
+        RailSwitchLeverComponent rslComponent = entity.getComponent(RailSwitchLeverComponent.class);
+        Vector3i blockLocation = new Vector3i(entity.getComponent(BlockComponent.class).getPosition());
```

Comment:
If you included the components in the method signature you don't have to use get component later.
Ie.

    @ReceiveEvent
    public void railSwitchLeverAction(ActivateEvent event, EntityRef entity, BlockComponent blockComponent, RailSwitchLeverComponent rslComponent)
    Vector3i blockLocation = new Vector3i(blockComponent.getPosition());